<a href="https://colab.research.google.com/github/tjwei/DataScienceSoftware2026/blob/main/pandas01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [141]:
import pandas as pd

In [142]:
pd.read_html('https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population')

HTTPError: HTTP Error 403: Forbidden

In [140]:
!curl -JLO https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1352k  100 1352k    0     0  13.8M      0 --:--:-- --:--:-- --:--:-- 13.9M


In [143]:
tables = pd.read_html('List_of_countries_and_dependencies_by_population')
len(tables)

3

### 解決 403 錯誤：加入 User-Agent
透過 `storage_options` 參數，我們可以自訂 HTTP 標頭（Headers），讓網站認為這是一般瀏覽器發出的請求。

In [144]:
import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population'

headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}
tables = pd.read_html(url, storage_options=headers)

print(len(tables))

df = tables[0]
display(df.head())

3


,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
0,World,8.232000e+09,100%,13 Jun 2025,UN projection[1][3],NaN
1,India,1.429404e+09,17.3%,1 Jul 2026,Official projection[4],[b]
2,China,1.404890e+09,17.0%,31 Dec 2025,Official estimate[5],[c]
3,United States,3.417849e+08,4.1%,1 Jul 2025,Official estimate[6],[d]
4,Indonesia,2.883151e+08,3.5%,31 Dec 2025,National annual projection[7],NaN


In [ ]:
df.shape

(241, 6)

In [ ]:
df.dtypes

,0
Location,object
Population,float64
% of world,object
Date,object
Source (official or from the United Nations),object
Notes,object


In [ ]:
df

,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
0,World,8.232000e+09,100%,13 Jun 2025,UN projection[1][3],NaN
1,India,1.429404e+09,17.3%,1 Jul 2026,Official projection[4],[b]
2,China,1.404890e+09,17.0%,31 Dec 2025,Official estimate[5],[c]
3,United States,3.417849e+08,4.1%,1 Jul 2025,Official estimate[6],[d]
4,Indonesia,2.883151e+08,3.5%,31 Dec 2025,National annual projection[7],NaN
...,...,...,...,...,...,...
236,Christmas Island (Australia),1.692000e+03,0%,1 Jan 2021,2021 Census[247],NaN
237,Niue (New Zealand),1.681000e+03,0%,11 Nov 2022,2022 Census[248],NaN
238,Vatican City,8.820000e+02,0%,31 Dec 2024,Official figure[249],[ah]
239,Cocos (Keeling) Islands (Australia),5.930000e+02,0%,30 Jun 2020,2021 Census[250],NaN


In [ ]:
df = df.dropna(how='all')

In [ ]:
df.Population.isna().any()

np.False_

In [145]:
pd.__version__

'2.2.3'

In [146]:
!pip list|grep pandas

geopandas                             1.2.0
pandas                                2.2.3
pandas-datareader                     0.11.1
pandas-gbq                            0.30.0
pandas-stubs                          2.2.3.250527
sklearn-pandas                        2.2.0


In [147]:
df.shape

(241, 6)

In [ ]:
df.Location

,Location
0,World
1,India
2,China
3,United States
4,Indonesia
...,...
236,Christmas Island (Australia)
237,Niue (New Zealand)
238,Vatican City
239,Cocos (Keeling) Islands (Australia)


In [ ]:
pd.read_

'2.2.3'

In [ ]:
!pip list|grep pandas

geopandas                             1.2.0
pandas                                2.2.3
pandas-datareader                     0.11.1
pandas-gbq                            0.30.0
pandas-stubs                          2.2.3.250527
sklearn-pandas                        2.2.0


In [ ]:
df[df['Location'].str.startswith('Ta')]

,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
21,Tanzania,70038850.0,0.8%,1 Jul 2026,Official projection[25],[h]
60,Taiwan,23224721.0,0.3%,31 Aug 2026,Official estimate[72],[p]
89,Tajikistan,10721000.0,0.1%,1 Jan 2026,Official estimate[100],NaN


In [ ]:
import humanize
humanize.number.intcomma(
    df.loc[1:].Population.mean()
    )

'33,780,573.43096235'

In [ ]:
df[1:].Population.quantile([0.25,0.5,0.75, 0.8])

In [ ]:
from matplotlib import pyplot as plt

In [ ]:
plt.hist(df[1:].Population, bins=100);

In [ ]:
df[df.Population > 100_000_000]

In [ ]:
df[df.Population > 1000_000_000].Population[1:].sum() / df.Population[0]

In [ ]:
type(df.Population)

In [ ]:
df['Location'].tail(20)

In [ ]:
df.loc[0]

In [ ]:
df.loc[0].index

In [ ]:
df['Population']

In [ ]:
df.T

In [ ]:
(df.T)[0]

In [ ]:
df[['Population', 'Population', 'Location', ]]

In [ ]:
df.loc[[100,11,1,236]]

In [148]:
import numpy as np
s = pd.Series(np.array([1, 3, 5, np.nan, 6, 8]))
s

,0
0,1.0
1,3.0
2,5.0
3,NaN
4,6.0
5,8.0


In [149]:
pd.Series(np.random.randn(5), index=['a', 'b', 'c', 'd', 'e'])

,0
a,-2.346758
b,-0.298317
c,0.749410
d,-1.113903
e,-2.003869


In [150]:
import numpy as np
import pandas as pd

degrees = [30, 45, 90]
sine_values = np.sin(np.deg2rad(degrees))

sine_table = pd.Series(sine_values, index=degrees)
sine_table

,0
30,0.500000
45,0.707107
90,1.000000


In [151]:
import numpy as np
import pandas as pd

# Adding more common angles
degrees = [0, 30, 45, 60, 90, 180, 270, 360]
sine_values = np.sin(np.deg2rad(degrees))

# Clean up floating point precision issues (e.g. sin(180) should be 0)
sine_values = np.round(sine_values, 10)

sine_table_expanded = pd.Series(sine_values, index=degrees)
sine_table_expanded

,0
0,0.000000
30,0.500000
45,0.707107
60,0.866025
90,1.000000
180,0.000000
270,-1.000000
360,-0.000000


In [166]:
dates = pd.date_range("20261007", periods=6)
dates

DatetimeIndex(['2026-10-07', '2026-10-08', '2026-10-09', '2026-10-10',
               '2026-10-11', '2026-10-12'],
              dtype='datetime64[ns]', freq='D')

In [170]:
df = pd.DataFrame(np.random.randn(6, 4),
                  index=dates.values,
                  columns=['a','b','c','d']
)

In [171]:
df

,a,b,c,d
2026-10-07,0.960356,0.881647,-0.104359,-1.459206
2026-10-08,-0.070537,1.017140,-0.267356,1.706352
2026-10-09,-0.339298,-0.062049,-1.821281,0.386062
2026-10-10,-1.205776,-0.400821,0.490747,-0.376372
2026-10-11,-0.115407,-0.166723,-0.834392,-0.478093
2026-10-12,0.662768,-0.036966,0.990115,0.336063


In [172]:
df2 = pd.DataFrame(
    {
        "A": 1.0,
        "B": pd.Timestamp("20130102"),
        "C": pd.Series(1, index=list(range(4)), dtype="float32"),
        "D": np.array([3] * 4, dtype="int32"),
        "E": pd.Categorical(["test", "train", "test", "train"]),
        "F": "foo",
    }
)


In [180]:
lut = pd.Series(["train", "test", "val"])
v= pd.Series([0,1,1,2,0,1,0,2])
lut[v]

,0
0,train
1,test
1,test
2,val
0,train
1,test
0,train
2,val


In [174]:
df2.dtypes

,0
A,float64
B,datetime64[s]
C,float32
D,int32
E,category
F,object


In [183]:
df2.T

,0,1,2,3
A,1.0,1.0,1.0,1.0
B,2013-01-02 00:00:00,2013-01-02 00:00:00,2013-01-02 00:00:00,2013-01-02 00:00:00
C,1.0,1.0,1.0,1.0
D,3,3,3,3
E,test,train,test,train
F,foo,foo,foo,foo


In [185]:
df2.T.dtypes

,0
0,object
1,object
2,object
3,object


In [186]:
import pandas as pd

# Create a simple dictionary
data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 35],
    'City': ['New York', 'London', 'Paris']
}

# Construct DataFrame using from_dict
df_dict = pd.DataFrame.from_dict(data)
df_dict

,Name,Age,City
0,Alice,25,New York
1,Bob,30,London
2,Charlie,35,Paris


In [ ]:
import pandas as pd

# Create a dictionary where we want the keys to be rows
data = {
    'Row_A': [1, 'a', 3],
    'Row_B': [4, 'b', 6],
    'Row_C': [7, 'c', 9]
}

# Construct DataFrame using from_dict with orient='index'
df_orient = pd.DataFrame.from_dict(data, orient='index', columns=['Col_1', 'Col_2', 'Col_3'])
df_orient

,Col_1,Col_2,Col_3
Row_A,1,a,3
Row_B,4,b,6
Row_C,7,c,9


In [187]:
import pandas as pd

# Create a list of tuples representing records
records = [
    (1, 'Alice', 'Engineer'),
    (2, 'Bob', 'Designer'),
    (3, 'Charlie', 'Artist')
]

# Construct DataFrame using from_records and specify column names
df_records = pd.DataFrame.from_records(records, columns=['ID', 'Name', 'Role'])
df_records

,ID,Name,Role
0,1,Alice,Engineer
1,2,Bob,Designer
2,3,Charlie,Artist


In [188]:
df.to_numpy()

array([[ 0.96035582,  0.88164718, -0.10435888, -1.45920578],
       [-0.07053715,  1.01713988, -0.26735553,  1.70635209],
       [-0.33929781, -0.06204859, -1.82128051,  0.38606197],
       [-1.20577581, -0.40082147,  0.49074732, -0.37637244],
       [-0.11540689, -0.16672334, -0.834392  , -0.47809299],
       [ 0.66276754, -0.0369662 ,  0.99011466,  0.33606304]])

In [189]:
df.values

array([[ 0.96035582,  0.88164718, -0.10435888, -1.45920578],
       [-0.07053715,  1.01713988, -0.26735553,  1.70635209],
       [-0.33929781, -0.06204859, -1.82128051,  0.38606197],
       [-1.20577581, -0.40082147,  0.49074732, -0.37637244],
       [-0.11540689, -0.16672334, -0.834392  , -0.47809299],
       [ 0.66276754, -0.0369662 ,  0.99011466,  0.33606304]])

In [190]:
df = pd.DataFrame(np.random.randn(6, 4), index=dates.values, columns=['a','b','c','d'])
df

,a,b,c,d
2026-10-07,-1.057073,-0.415894,-0.332035,0.564773
2026-10-08,0.424437,-1.286526,0.808120,1.299867
2026-10-09,-0.924757,0.564242,-0.333730,1.142868
2026-10-10,0.121268,0.523218,0.769006,-1.236708
2026-10-11,0.122820,0.393334,-0.912020,-1.430416
2026-10-12,-0.381725,-0.488556,0.528373,-0.067912


In [ ]:
df.loc['2026-10-8':'2026-10-11']

,a,b,c,d
2026-10-08,-0.007596,-1.101149,-0.688045,-0.249796
2026-10-09,-0.965306,-0.484330,-0.525757,-0.379421
2026-10-10,0.096305,-1.136206,0.804640,-0.555662
2026-10-11,0.140729,1.326791,1.112832,-0.494450


In [ ]:
df2 = df.loc[['2026-10-12','2026-10-9','2026-10-07','2026-10-8','2026-10-11']].copy()
df2

,a,b,c,d
2026-10-12,-0.940473,0.643538,-0.503280,0.009432
2026-10-09,-0.965306,-0.484330,-0.525757,-0.379421
2026-10-07,-0.873419,-0.238330,-0.350945,1.854248
2026-10-08,-0.007596,-1.101149,-0.688045,-0.249796
2026-10-11,0.140729,1.326791,1.112832,-0.494450


In [ ]:
df2.loc['2026-10-09':]

,a,b,c,d
2026-10-12,-0.940473,0.643538,-0.503280,0.009432
2026-10-09,-0.965306,-0.484330,-0.525757,-0.379421
2026-10-11,0.140729,1.326791,1.112832,-0.494450


In [210]:
df2 = df.copy()
df2['G'] = pd.Series([10,20,30,40,50,60], index=df2.index)
df2

,a,b,c,d,F,G
2026-10-07,-1.057073,-0.415894,-0.332035,0.564773,1,10
2026-10-08,0.424437,-1.286526,0.808120,1.299867,2,20
2026-10-09,-0.924757,0.564242,-0.333730,1.142868,3,30
2026-10-10,0.121268,0.523218,0.769006,-1.236708,4,40
2026-10-11,0.122820,0.393334,-0.912020,-1.430416,5,50
2026-10-12,-0.381725,-0.488556,0.528373,-0.067912,6,60


In [207]:
pd.Series([10,20,30,40,50,60])

,0
0,10
1,20
2,30
3,40
4,50
5,60


In [219]:
df2 = df.copy()
new_F = pd.Series([100,200], index=[dates[3], dates[1]])
df2['G'] = new_F
df2.loc[~df2['G'].isna(), 'F']=new_F
del df2['G']
df2


,a,b,c,d,F
2026-10-07,-1.057073,-0.415894,-0.332035,0.564773,1
2026-10-08,0.424437,-1.286526,0.808120,1.299867,200
2026-10-09,-0.924757,0.564242,-0.333730,1.142868,3
2026-10-10,0.121268,0.523218,0.769006,-1.236708,100
2026-10-11,0.122820,0.393334,-0.912020,-1.430416,5
2026-10-12,-0.381725,-0.488556,0.528373,-0.067912,6


In [220]:
df2 = df.copy()
new_F = pd.Series([100,200], index=[dates[3], dates[1]])
df2.loc[new_F.index, 'F'] = new_F
df2


,a,b,c,d,F
2026-10-07,-1.057073,-0.415894,-0.332035,0.564773,1
2026-10-08,0.424437,-1.286526,0.808120,1.299867,200
2026-10-09,-0.924757,0.564242,-0.333730,1.142868,3
2026-10-10,0.121268,0.523218,0.769006,-1.236708,100
2026-10-11,0.122820,0.393334,-0.912020,-1.430416,5
2026-10-12,-0.381725,-0.488556,0.528373,-0.067912,6


In [227]:
df2 = df.copy()
new_F = pd.Series([100,200], index=[dates[3], dates[1]])
new_F = new_F.reindex(index=df2.index)
df2.loc[~new_F.isna(), 'F'] = new_F
df2


,a,b,c,d,F
2026-10-07,-1.057073,-0.415894,-0.332035,0.564773,1
2026-10-08,0.424437,-1.286526,0.808120,1.299867,200
2026-10-09,-0.924757,0.564242,-0.333730,1.142868,3
2026-10-10,0.121268,0.523218,0.769006,-1.236708,100
2026-10-11,0.122820,0.393334,-0.912020,-1.430416,5
2026-10-12,-0.381725,-0.488556,0.528373,-0.067912,6


In [ ]:
pd.Series([1,2,3,4,5,6])

,0
0,1
1,2
2,3
3,4
4,5
5,6


In [191]:
df2 = df.copy()
df2["E"] = ["one", "one", "two", "three", "four", "three"]
df2

,a,b,c,d,E
2026-10-07,-1.057073,-0.415894,-0.332035,0.564773,one
2026-10-08,0.424437,-1.286526,0.808120,1.299867,one
2026-10-09,-0.924757,0.564242,-0.333730,1.142868,two
2026-10-10,0.121268,0.523218,0.769006,-1.236708,three
2026-10-11,0.122820,0.393334,-0.912020,-1.430416,four
2026-10-12,-0.381725,-0.488556,0.528373,-0.067912,three


In [194]:
df[df2['E'].apply(lambda x: x in ["two", "four"])]

,a,b,c,d
2026-10-09,-0.924757,0.564242,-0.33373,1.142868
2026-10-11,0.122820,0.393334,-0.91202,-1.430416


In [229]:
df1 = pd.DataFrame({'A':[1], 'B':[2], 'C':[3]})
df2 = pd.DataFrame({'A':[30,40], 'D':[90,80]}, index=[0,1])
df1+df2


,A,B,C,D
0,31.0,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN


In [230]:
s=pd.Series(list("jdsailkgjlkdjkfjdsajfijeafjjfskdaljfasjijskdsvna"))
s

,0
0,j
1,d
2,s
3,a
4,i
5,l
6,k
7,g
8,j
9,l


In [236]:
s.value_counts()

,count
j,11
s,6
a,6
d,5
k,5
f,5
l,3
i,3
g,1
e,1


In [256]:
import pandas as pd

# DataFrame 1: Name and Age
df_age = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'Charlie', 'Alice'],
    'Age': [25, 30, 35, 40]
})

# DataFrame 2: Name and Height
df_height = pd.DataFrame({
    'Name': ['Bob', 'Charlie','Alice',  'Alice'],
    'Height': [165, 180, 175, 160]
})

print("--- DataFrame 1: Age ---")
display(df_age)
print("\n--- DataFrame 2: Height ---")
display(df_height)

--- DataFrame 1: Age ---


,Name,Age
0,Alice,25
1,Bob,30
2,John,35
3,David,40



--- DataFrame 2: Height ---


,Name,Height
0,Bob,165
1,Charlie,180
2,Alice,175
3,Eve,160


In [255]:
pd.merge(df_age, df_height, on='Name', how='left')

,Name,Age,Height
0,Alice,25,175
1,Alice,25,160
2,Bob,30,165
3,Charlie,35,180
4,Alice,40,175
5,Alice,40,160


In [259]:
import pandas as pd

# DataFrame 1: Name and Age
df_age = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'Charlie', 'Davide'],
    'Age': [25, 30, 35, 40]
})
df_age.set_index('Name', inplace=True)
# DataFrame 2: Name and Height
df_height = pd.DataFrame({
    'Name': ['Bob', 'Charlie','Eve',  'John'],
    'Height': [165, 180, 175, 160]
})
df_height.set_index('Name', inplace=True)

print("--- DataFrame 1: Age ---")
display(df_age)
print("\n--- DataFrame 2: Height ---")
display(df_height)

--- DataFrame 1: Age ---


,Age
Name,
Alice,25
Bob,30
Charlie,35
Davide,40



--- DataFrame 2: Height ---


,Height
Name,
Bob,165
Charlie,180
Eve,175
John,160


In [263]:
pd.concat([df_age, df_height])

,Age,Height
0,25.0,NaN
1,30.0,NaN
2,35.0,NaN
3,40.0,NaN
4,NaN,165.0
5,NaN,180.0
6,NaN,175.0
7,NaN,160.0
